File: libs/shared-models/src/link.ts
import { Coordinate, PaginationRequest, PaginationResponse } from './common';
import { EntityType } from './entity';

// Canonical definition in @sentinel/proto-gen (proto/link.proto)
export enum LinkType {
  COMMUNICATION = 'COMMUNICATION',
  ASSOCIATION = 'ASSOCIATION',
  COLOCATION = 'COLOCATION',
  FINANCIAL = 'FINANCIAL',
  ORGANIZATIONAL = 'ORGANIZATIONAL',
  MOVEMENT_PATTERN = 'MOVEMENT_PATTERN',
  COMMAND_CONTROL = 'COMMAND_CONTROL',
  GEOGRAPHIC = 'GEOGRAPHIC',
  FAMILIAL = 'FAMILIAL',
  LOGISTIC = 'LOGISTIC',
  OPERATIONAL = 'OPERATIONAL',
  IDENTITY = 'IDENTITY',
}

export enum LinkEventType {
  CREATED = 'CREATED',
  UPDATED = 'UPDATED',
  DELETED = 'DELETED',
}

export interface Link {
  linkId: string;
  sourceEntityId: string;
  targetEntityId: string;
  linkType: LinkType;
  confidence: number;
  description?: string;
  evidence: string[];
  firstObserved?: string;
  lastObserved?: string;
  metadata: Record<string, string>;
}

export interface GraphNode {
  entityId: string;
  entityType: EntityType;
  name: string;
  position?: Coordinate;
}

export interface GraphEdge {
  link: Link;
}

export interface GetLinksRequest extends PaginationRequest {
  entityId: string;
  linkTypes?: LinkType[];
  minConfidence?: number;
  maxHops?: number;
}

export interface GetLinksResponse {
  links: Link[];
  pagination: PaginationResponse;
}

export interface GetGraphRequest {
  centerEntityId: string;
  maxDepth: number;
  linkTypes?: LinkType[];
  minConfidence?: number;
}

export interface GetGraphResponse {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

File: libs/common-go/models/link.go
package models

// LinkType constants.
const (
	LinkTypeCommunication   = "COMMUNICATION"
	LinkTypeAssociation     = "ASSOCIATION"
	LinkTypeColocation      = "COLOCATION"
	LinkTypeFinancial       = "FINANCIAL"
	LinkTypeOrganizational  = "ORGANIZATIONAL"
	LinkTypeMovementPattern = "MOVEMENT_PATTERN"
	LinkTypeCommandControl  = "COMMAND_CONTROL"
	LinkTypeGeographic      = "GEOGRAPHIC"
	LinkTypeFamilial        = "FAMILIAL"
	LinkTypeLogistic        = "LOGISTIC"
	LinkTypeOperational     = "OPERATIONAL"
	LinkTypeIdentity        = "IDENTITY"
)

// Link represents a relationship between two entities.
type Link struct {
	LinkID         string            `json:"linkId"`
	SourceEntityID string            `json:"sourceEntityId"`
	TargetEntityID string            `json:"targetEntityId"`
	LinkType       string            `json:"linkType"`
	Confidence     float64           `json:"confidence"`
	Description    *string           `json:"description,omitempty"`
	Evidence       []string          `json:"evidence"`
	FirstObserved  *string           `json:"firstObserved,omitempty"`
	LastObserved   *string           `json:"lastObserved,omitempty"`
	Metadata       map[string]string `json:"metadata"`
}

// GraphNode represents a node in the entity graph.
type GraphNode struct {
	EntityID   string      `json:"entityId"`
	EntityType string      `json:"entityType"`
	Name       string      `json:"name"`
	Position   *Coordinate `json:"position,omitempty"`
}

// GraphEdge represents an edge in the entity graph.
type GraphEdge struct {
	Link Link `json:"link"`
}
